src/client/Microsoft.Identity.Client/ManagedIdentity/MachineLearningManagedIdentitySource.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Globalization;
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.Internal;
+
+namespace Microsoft.Identity.Client.ManagedIdentity
+{
+    internal class MachineLearningManagedIdentitySource : AbstractManagedIdentity
+    {
+        private const string MachineLearningMsiApiVersion = "2017-09-01";
+        private const string SecretHeaderName = "secret";
+
+        private readonly Uri _endpoint;
+        private readonly string _secret;
+
+        public static AbstractManagedIdentity Create(RequestContext requestContext)
+        {
+            requestContext.Logger.Info(() => "[Managed Identity] Machine learning managed identity is available.");
+
+            return TryValidateEnvVars(EnvironmentVariables.MsiEndpoint, requestContext.Logger, out Uri endpointUri)
+                ? new MachineLearningManagedIdentitySource(requestContext, endpointUri, EnvironmentVariables.MsiSecret)
+                : null;
+        }
+
+        private MachineLearningManagedIdentitySource(RequestContext requestContext, Uri endpoint, string secret) 
+            : base(requestContext, ManagedIdentitySource.MachineLearning)
+        {
+            _endpoint = endpoint;
+            _secret = secret;
+        }
+
+        private static bool TryValidateEnvVars(string msiEndpoint, ILoggerAdapter logger, out Uri endpointUri)
```

Comment:
nit: It looks like this method only returns `true` or throws. Should it be changed to something like `private static Uri ValidateEnvVarsOrThrow(string msiEndpoint, ILoggerAdapter logger);`

Reply:
sure we can explore, there are few other sources this may need a change if we do so. Let me file a work item for this to track